### PR TITLE
Document fli instruction from zfa extension

### PR DESCRIPTION
**Pull request by kito-cheng**
Split non-zfa part into https://github.com/riscv-non-isa/riscv-asm-manual/pull/86
This is chaining PR with #86

---

Zfa has provide fli instruction to load a floating point immediate, but some of value are hard to written by human, so we would like to support 3 different format of floating value: hexadecimal, decimal, or standard notation formats (a.k.a scientific notation); document this could let user easier know how to use those value in their asm code.

### CODE REVIEW

**Comment by kito-cheng** [comment]
@asb @cmuellner @topperc @a4lg

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:593]: diretive -> directive

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:594]: float -> floating

Comment (cmuellner) [reply riscv-asm.md:594]: and the load into -> and then load it into a

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:605]: I would remove this comment (and optionally wait for Zfa to be ratified).

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:584]: What about `62.5e-3`?

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:537]: Do we need to duplicate the list here (i.e. can we not refer to the specification)?

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:534]: Does this need to be Zfa-specific?

We could instead define a generic way of supported floating-point literal formats supported by the assembler.
Example uses of such floating-point literals would be in `.single <expr>` and `fli.s fN, <expr>`.

The restriction of supported values for `fli` is a restriction that is instruction specific and does not need to be documented here.

**Review by cmuellner — COMMENTED** [review]
Thanks for the proposal.
I think it is a good idea to document the supported floating-point literal formats that are supported by the assembler.

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:593]: Damm, I just forgot remove the original version, plz just ignore this paragraph.

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:594]: Same here

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:534]: Yeah, I thought same thing, that will require one extra temporal register to hold the address, but that should not really big deal.

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:537]: I would like to hold the this here to let user has a way to quick look up how to write those value in any format (he like), since ISA spec isn't provide this table.

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:605]: Hmm, sounds like good idea and I could split non-zfa part into a separated PR.

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:584]: Tried on both LLVM and binutils, they accept that as well, but I guess we don't need to document that here since it's kind of standard way?

Comment (cmuellner) [reply riscv-asm.md:584]: With this question, I wanted to show that an enumeration of valid values is not optimal.

Instead we could reference the floating-point literal formats that are supported. A good overview can be found here: https://en.cppreference.com/w/cpp/language/floating_literal.

File: riscv-asm.md
Comment (cmuellner) [inline riscv-asm.md:578]: Is your "decimal" representation not a special case of the "standard" notation?

When looking into `man strtof` I only see a decimal format and a hexadecimal format.
E.g. also `-1.` and `-0x1.` are valid floating point literals, but would not fit any format of your table.

File: riscv-asm.md
Comment (aswaterman) [inline riscv-asm.md:543]: Re: `1.525879e-05`: I thought we had agreed to disallow inexact constants, so that this variant would be rejected but `1.52587890625e-06` would be accepted?

**Comment by kito-cheng** [comment]
Split non-zfa part into #86

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:543]: My bad, this table is generated from my C code, I did check output for decimal but not well check on Scientific Notation form

Comment (aswaterman) [reply riscv-asm.md:543]: Cool.

File: riscv-asm.md
Comment (topperc) [inline riscv-asm.md:602]: is flh correct for loading from .float?

Comment (kito-cheng) [reply riscv-asm.md:602]: thanks, that should be flw

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md:580]: Added a note, this should improved:
```
NOTE: A value can be expressed in different forms within the same format.
For example, 6.5536e+04 can be written as 6553.6e+01 or 65.536e+03. The table
lists one possible representation, but any exact value can be used.
```

File: riscv-asm.md
Comment (kito-cheng) [reply riscv-asm.md]: reoreded the paragraph, reference to RISC-V ISA spec instead, but put the table in the button of this section and mark as NOTE.

**Comment by kito-cheng** [comment]
Changes:
- Split non-zfa part into https://github.com/riscv-non-isa/riscv-asm-manual/pull/86
- Add nominate text to say tool should reject inexact value.
- Add note to say a value written in any form that has any exact value (e.g. 6.5536e+04 can be written as 6553.6e+01 or 65.536e+03)
- Add https://en.cppreference.com/w/cpp/language/float this as ref.

**Comment by a4lg** [comment]
Thanks for proposal @kito-cheng!

It's good to hear that the constant operand must be exact (if not, I would be worried about its precision for fuzzy matching).

**Comment by kito-cheng** [comment]
Changes:
- Reference to `Floating-point literals` instead of reference to something like hexadecimal.
- Revise the wording.

**Comment by asb** [comment]
I think it's helpful to have this documented somewhere, but defer to others on detailed line-by-line correctness review.

File: riscv-asm.md
Comment (asb) [inline riscv-asm.md:548]: instruction => instructions

File: riscv-asm.md
Comment (asb) [inline riscv-asm.md:549]: value => values. Though we list all the values below, so perhaps "supported values can be found in the RISC-V ISA specification but are also listed below."

**Comment by kito-cheng** [comment]
@asb thanks, suggestion applied :)

**Comment by kito-cheng** [comment]
ping :)

File: riscv-asm.md
Comment (aswaterman) [inline riscv-asm.md:591]: I suggest changing "Possible inputs values" to "Example legal input values".  This makes it clear these aren't the only possibilities.  (Also note that "input" should not be plural.)

**Review by aswaterman — APPROVED** [review]
LGTM modulo my last suggestion.

**Review by cmuellner — APPROVED** [review]
LGTM (after applying the change requested by @aswaterman).

**Comment by kito-cheng** [comment]
Updated, thanks :)

**Pull request merged**
